File: app/src/main/java/com/example/remindo/database/RemindoFireBase.java
package com.example.remindo.database;

import android.os.Parcel;
import android.os.Parcelable;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.example.remindo.Adapter.RemindoAdapter;
import com.example.remindo.Models.RemindoModel;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentChange;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.QuerySnapshot;
import java.util.ArrayList;
import java.util.List;

public class RemindoFireBase implements Parcelable {
    String TAG = "FireBase";
    FirebaseFirestore db = FirebaseFirestore.getInstance();
    CollectionReference toDoCollectionReference;
    RemindoAdapter adapter;
    List<DocumentSnapshot> documentSnapshotList = new ArrayList<>();

    public RemindoFireBase(RemindoAdapter adapter) {
        this.adapter = adapter;
        toDoCollectionReference = db.collection("Users")
                .document("1")
                .collection("ToDo");

        toDoCollectionReference.addSnapshotListener(new EventListener<QuerySnapshot>() {
            @Override
            public void onEvent(@Nullable QuerySnapshot value, @Nullable FirebaseFirestoreException error) {
                for (DocumentChange dc : value.getDocumentChanges()) {
                    switch (dc.getType()) {
                        case MODIFIED:
                        case REMOVED:
                            readFromFireBase();
                            break;
                    }
                }
            }
        });


    }

    protected RemindoFireBase(Parcel in) {
        TAG = in.readString();
    }

    public static final Creator<RemindoFireBase> CREATOR = new Creator<RemindoFireBase>() {
        @Override
        public RemindoFireBase createFromParcel(Parcel in) {
            return new RemindoFireBase(in);
        }

        @Override
        public RemindoFireBase[] newArray(int size) {
            return new RemindoFireBase[size];
        }
    };

    public RemindoFireBase() {

    }

    public void addToFireBase(RemindoModel rvm){
        db.collection("Users")
                .document("1")
                .collection("ToDo")
                 .add(rvm)
                 .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
                    @Override
                    public void onSuccess(DocumentReference documentReference) {
                        Log.d(TAG, "DocumentSnapshot added with ID: " + documentReference.getId());
                    }
                }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        System.out.println(e.getMessage());
                    }
                });
    }

    public void readFromFireBase(){
        db.collection("Users")
                .document("1")
                .collection("ToDo")
                .orderBy("done").orderBy("priority")
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                       documentSnapshotList = task.getResult().getDocuments();
                       adapter.notifyDataSetChanged();
                    }
                });
    }


    public void updateFireBaseData(String id, RemindoModel rvm){
        db.collection("Users")
                .document("1")
                .collection("ToDo")
                .document(id)
                .set(rvm)
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void unused) {
                        System.out.println("Done Update Sucess " + id + "  " + rvm.getDone());
                    }
                });
    }
    public DocumentSnapshot getDocumentData(int pos){
        return documentSnapshotList.get(pos);
    }

    public int getSize(){
        return documentSnapshotList.size();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeString(TAG);
    }
}

File: app/src/main/java/com/example/remindo_kot/Database/RemindoFireBase.kt
package com.example.remindo_kot.Database

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.util.Log
import android.widget.Toast
import com.example.remindo_kot.Adapter.RemindoAdapter
import com.example.remindo_kot.Models.RemindoModel
import com.google.android.gms.tasks.OnFailureListener
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.DocumentChange
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase

class RemindoFireBase() : Parcelable {
    val db = Firebase.firestore
    lateinit var toDoCollectionReference :CollectionReference
    lateinit var adapter:RemindoAdapter
    var documentSnapshotList: List<DocumentSnapshot> = ArrayList()

    constructor(parcel: Parcel) : this() {
    }

    constructor(adapter: RemindoAdapter) : this() {
        this.adapter = adapter
        toDoCollectionReference = db.collection("Users")
            .document("1")
            .collection("ToDo")
        toDoCollectionReference.addSnapshotListener { value, error ->
            for (dc in value!!.documentChanges) {
                when (dc.type) {
                    DocumentChange.Type.MODIFIED, DocumentChange.Type.REMOVED -> readFromFireBase()
                    else -> {break}

                }
            }
        }
    }

fun addToFireBase(rvm: RemindoModel,context: Context){
    db.collection("Users")
        .document("1")
        .collection("ToDo")
        .add(rvm)
        .addOnSuccessListener { documentReference ->
            Toast.makeText(context,"Task Added",Toast.LENGTH_SHORT).show()
            Log.d(CREATOR.TAG, "DocumentSnapshot added with ID: " + documentReference.id)
        }.addOnFailureListener(OnFailureListener { e -> println(e.message) })
}
    fun readFromFireBase(){
        db.collection("Users")
            .document("1")
            .collection("ToDo")
            .orderBy("done").orderBy("priority")
            .get()
            .addOnCompleteListener{task ->
                documentSnapshotList = task.getResult().documents
                adapter.notifyDataSetChanged()
            }
    }

    fun updateFireBaseData(id:String,rvm:RemindoModel){
        db.collection("Users")
            .document("1")
            .collection("ToDo")
            .document(id)
            .set(rvm)
            .addOnSuccessListener{unused ->
                println("Done Update Sucess " + id + "  " + rvm.done)
            }
    }

    fun getDocumentData(pos:Int): DocumentSnapshot {
        return documentSnapshotList.get(pos)
    }

    fun getSize(): Int {
        return documentSnapshotList.size
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {

    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<RemindoFireBase> {
        override fun createFromParcel(parcel: Parcel): RemindoFireBase {
            return RemindoFireBase(parcel)
        }

        override fun newArray(size: Int): Array<RemindoFireBase?> {
            return arrayOfNulls(size)
        }

        private const val TAG = "RemindoFireBase"
    }


}